internal/mempool/mempool_test.go

```diff
@@ -3189,3 +3189,63 @@ func TestMiningView(t *testing.T) {
 			"even though it was removed from the mempool.")
 	}
 }
+
+// TestAncestorTrackingLimits ensures that ancestor tracking is
+// disabled for transactions that have too many ancestors.
+func TestAncestorTrackingLimits(t *testing.T) {
+	harness, spendableOuts, err := newPoolHarness(chaincfg.MainNetParams())
+	if err != nil {
+		t.Fatalf("unable to create test pool: %v", err)
+	}
+
+	// Add a chain of transactions to the mempool.
+	var allTxns []*dcrutil.Tx
+	prevSpendableOut := spendableOuts[0]
+	for i := 0; i < ancestorTrackingLimit+1; i++ {
+		tx, _ := harness.CreateSignedTx([]spendableOutput{
+			prevSpendableOut,
+		}, 1)
+
+		allTxns = append(allTxns, tx)
+		prevSpendableOut = txOutToSpendableOut(tx, 0, wire.TxTreeRegular)
+	}
+
+	// Add all transactions to the mempool.
+	for index, tx := range allTxns {
+		_, err := harness.txPool.ProcessTransaction(tx,
+			true, false, true, 0)
+		if err != nil {
+			t.Fatalf("ProcessTransaction: failed to accept valid "+
+				"transaction at index %d: %v", index, err)
+		}
+	}
+
+	// Remove each transaction from the mempool to demonstrate that once a
+	// transaction has less ancestors than the limit that it begins having
+	// ancestor stats tracked and it continues to do so until it is removed
+	// from the mempool.
+	for len(allTxns) > 0 {
+		for index, tx := range allTxns {
+			_, hasStats := harness.txPool.miningView.AncestorStats(tx.Hash())
+			if index < ancestorTrackingLimit && !hasStats {
```

Comment:
Should this be `index <= ancestorTrackingLimit`?

Reply:
Yes, this has been fixed. Thank you. On a related note, there was an off-by-one bug here where a transaction being removed from the graph was counted as an ancestor, which prevented a descendant from having ancestor statistics tracked due to having one-too-many ancestors.